client/pages/discover-page.jsx

```diff
@@ -0,0 +1,138 @@
+import React from 'react';
+import Form from 'react-bootstrap/Form';
+import Button from 'react-bootstrap/Button';
+import Redirect from '../components/redirect';
+import SearchResult from '../components/search-result';
+import AppContext from '../lib/app-context';
+
+export default class Discover extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      searchInput: ''
+    };
+
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleFilterClick = this.handleFilterClick.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const { params } = this.context.route;
+    const endpoint = params.get('filterType');
+    const prevEndpoint = prevProps.route.params.get('filterType');
+    const q = this.state.searchInput;
+    if (!q) {
+      // eslint-disable-next-line no-useless-return
+      return;
+    } else if (prevState.searchInput !== q || prevEndpoint !== endpoint) {
+      fetch(`/api/search/${endpoint}?q=${q}`)
+        .then(res => res.json())
+        .then(result => {
+          this.setState({ result });
+        });
+    }
+  }
+
+  handleChange(e) {
+    const { value } = e.target;
+    this.setState({ searchInput: value });
+  }
+
+  handleSubmit(e) {
+    e.preventDefault();
+    const { params } = this.context.route;
+    let endpoint = params.get('filterType');
+    const q = this.state.searchInput;
+    if (!endpoint) {
+      window.location.hash = '#discover?filterType=track';
+      endpoint = 'track';
+    }
+    fetch(`/api/search/${endpoint}?q=${q}`)
+      .then(res => res.json())
+      .then(result => {
+        this.setState({ result });
+      });
+
+  }
```

Comment:
For `handleSubmit`, we should put the user's search in the hash too and not do the fetch here. Instead, we just update the hash so that `componentDidUpdate` sends the fetch

```js
 handleSubmit(e) {
    e.preventDefault();
    const { params } = this.context.route;
    let endpoint = params.get('filterType');
    const q = this.state.searchInput;
    let newHash = `#discover?`
    if (endpoint) {
      newHash += `filterType=${endpoint}`
    } else {
      newHash += `filterType=track`
    }
    if (q) {
      newHash += `&q=${q}`
    }
    window.location.hash = newHash;
  }
```